odl/trafos/util/ft_utils.py

```diff
@@ -0,0 +1,654 @@
+    Parameters
+    ----------
+    arr : `array-like`
+        Array to be pre-processed. An array with real data type is
+        converted to its complex counterpart.
+    real_grid : `RegularGrid`
+        Real space grid in the transform
+    recip_grid : `RegularGrid`
+        Reciprocal grid in the transform
+    shift : bool or `sequence` of bools
+        If ``True``, the grid is shifted by half a stride in the negative
+        direction in the corresponding axes. The sequence must have the
+        same length as ``axes``.
+    axes : int or `sequence` of ints
+        Dimensions along which to take the transform. The sequence must
+        have the same length as ``shifts``.
+    interp : string or `sequence` of strings
+        Interpolation scheme used in the real-space.
+    sign : {'-', '+'}, optional
+        Sign of the complex exponent.
+    op : {'multiply', 'divide'}
+        Operation to perform with the stride times the interpolation
+        kernel FT
+    out : `numpy.ndarray`, optional
+        Array in which the result is stored. If ``out is arr``, an
+        in-place modification is performed.
+
+    Returns
+    -------
+    out : `numpy.ndarray`
+        Result of the post-processing. If ``out`` was given, the returned
+        object is a reference to it.
+    """
+    arr = np.asarray(arr)
+    if is_real_floating_dtype(arr.dtype):
+        arr = arr.astype(TYPE_MAP_R2C[arr.dtype])
+    elif not is_complex_floating_dtype(arr.dtype):
+        raise ValueError('array data type {} is not a floating point data '
+                         'type'.format(dtype_repr(arr.dtype)))
+
+    if out is None:
+        out = arr.copy()
+    elif out is not arr:
+        out[:] = arr
+
+    if axes is None:
+        axes = list(range(arr.ndim))
+    else:
+        try:
+            axes = [int(axes)]
+        except TypeError:
+            axes = list(axes)
+
+    shift_list = normalized_scalar_param_list(shift, length=len(axes),
+                                              param_conv=bool)
+
+    if sign == '-':
+        imag = -1j
+    elif sign == '+':
+        imag = 1j
+    else:
+        raise ValueError("`sign` '{}' not understood".format(sign))
+
+    op, op_in = str(op).lower(), op
+    if op not in ('multiply', 'divide'):
+        raise ValueError("kernel `op` '{}' not understood".format(op_in))
+
+    # Make a list from interp if that's not the case already
+    try:
+        interp = [str(interp + '').lower()] * arr.ndim
+    except TypeError:
+        pass
+
+    onedim_arrs = []
+    for ax, shift, intp in zip(axes, shift_list, interp):
+        x = real_grid.min_pt[ax]
+        xi = recip_grid.coord_vectors[ax]
+
+        # First part: exponential array
+        onedim_arr = (np.exp(imag * x * xi))
+
+        # Second part: interpolation kernel
+        len_dft = recip_grid.shape[ax]
+        len_orig = real_grid.shape[ax]
+        halfcomplex = (len_dft < len_orig)
+        odd = len_orig % 2
+
+        if shift:
```

Comment:
This if if if construction is a mess to review. I frankly cant tell if it is correct or not. Is there no clearer way?